app/models/sign_in/concerns/certifiable.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+module SignIn
+  module Concerns
+    module Certifiable
+      extend ActiveSupport::Concern
+
+      def assertion_public_keys
+        @assertion_public_keys ||= certificate_objects.map(&:public_key)
+      end
+
+      def certificate_with_serial(serial)
+        certificate_objects.find do |certificate|
+          certificate.serial.to_s == serial
+        end
+      end
+
+      def delete_certificate_with_serial(serial)
+        original_length = certificate_objects.length
+
+        @certificate_objects = certificate_objects.delete_if do |certificate|
+          certificate.serial.to_s == serial
+        end
+
+        return false unless original_length > certificate_objects.length
+
+        update(certificates: certificate_objects.map(&:to_pem))
+      end
+
+      def expired_certificates
```

Comment:
I wonder if these functions should just be in the `CertificateCheckerJob`. I'm not sure if there is a need to have these functions available outside of that usage. My reasoning is that the ways we currently validate certs will already do some of this stuff (like `JWT.decode` validation), and we explicitly don't want anyone to use these functions directly for validation (since we prefer the 'official' ways to validate).

Reply:
I'm not sure if these will be directly useful on Rails Console, cause here's how I'm thinking the process will currently look like:
- Job says cert is about to expire
- new cert is procured
- Engineer takes public cert and adds it to config cert list (just append to array)
- (At some later date), job says cert has expired
- Engineer removes expired cert 

The 'Engineer removes expired cert` is where these functions could be used, but tbh it should probably be a rake task that's set up to remove the expired cert

Reply:
I think these approaches show a need for a complete certificate lifecycle to be considered, which is exactly why I proposed a new 'certificate management' project for us to work on hopefully pretty soon. But in the meantime I'm trying to think of MVP and what would be easiest to refactor into a better, holistic solution for everything from adding, monitoring, removing certs.

I think in the meantime having as little new functionality in the configuration classes as possible and everything in the Job is easiest? At least that's where my thinking is at

Reply:
I can do that, but I'm hesitant to throw all the logic into the job because it just won't be as ergonomic.

For instance, the current implementation I have allows us to do this:

```ruby
config.expired_certificates.each do |certificate|
  payload = payload_for_alert(config, certificate)
  Rails.logger.warn('[SignIn] [CertificateChecker] expired_certificate', payload)
end
```

Putting all the logic in the job would require something like this:

```ruby
certificate_objects = config.certificates.compact.map do |certificate|
  OpenSSL::X509::Certificate.new(certificate)
end

expired_certificates = certificate_objects.select do |certificate|
  certificate.not_after < Time.zone.now
end

expired_certificates.each do |expired_certificate|
  payload = payload_for_alert(config, expired_certificate)
  Rails.logger.warn('[SignIn] [CertificateChecker] expired_certificate', payload)
end
```

I feel like the model concern route sets us up nicely for a refactor as we'd only need to reimplement the `expired_certificates` method to be a scope that queries for associated certificate records that match the criteria. Ditto for `expiring_certificates` and `self_signed_certificates`. We'd end up trimming some other private methods from the concern as they'd no longer be needed. We could also include the relationship in the concern:

```ruby
included do
  has_many :sign_in_certificates, through: :configuration_certificates
end
```

Ultimately, I'll defer to you on this, but I really think a model concern is the way to go here.

Reply:
sounds good to me, just wanted to press on that to see what shakes out. 

It looks like you already removed the functions that the job doesn't use, so to me this looks good to go. I'll do a final test and approve soon

Reply:
Nice, thanks for forcing me to put my thoughts into words.